File: ontology-engine/graph-core/src/main/java/org/sunbird/graph/validator/NodeValidator.java
package org.sunbird.graph.validator;

import org.sunbird.common.exception.ResourceNotFoundException;
import org.sunbird.common.exception.ServerException;
import org.sunbird.graph.common.enums.SystemProperties;
import org.sunbird.graph.dac.model.Filter;
import org.sunbird.graph.dac.model.MetadataCriterion;
import org.sunbird.graph.dac.model.Node;
import org.sunbird.graph.dac.model.SearchConditions;
import org.sunbird.graph.dac.model.SearchCriteria;
import org.sunbird.graph.exception.GraphEngineErrorCodes;
import org.sunbird.graph.service.operation.SearchAsyncOperations;
import scala.compat.java8.FutureConverters;
import scala.concurrent.Future;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;

/**
 * This class provides utility methods for node validation
 * @author Kumar Gauraw
 */
public class NodeValidator {

    /**
     * This method validates whether given identifiers exist in the given graph or not.
     * @param graphId
     * @param identifiers
     * @return List<Node>
     */
    public static Future<Map<String, Node>> validate(String graphId, List<String> identifiers) {
        List<Map<String, Object>> result;
        Future<List<Node>> nodes = getDataNodes(graphId, identifiers);
        CompletionStage<Map<String, Node>> cs = FutureConverters.toJava(nodes).thenApply(dataNodes -> {
            Map<String, Node> relationNodes = new HashMap<>();
            if (dataNodes.size() != identifiers.size()) {
                List<String> invalidIds = identifiers.stream().filter(id -> dataNodes.stream().noneMatch(node -> node.getIdentifier().equals(id)))
                        .collect(Collectors.toList());
                throw new ResourceNotFoundException(GraphEngineErrorCodes.ERR_INVALID_NODE.name(), "Node Not Found With Identifier " + invalidIds);
            } else {
                relationNodes = dataNodes.stream().collect(Collectors.toMap(node -> node.getIdentifier(), node -> node));
                return relationNodes;
            }
        });

        return FutureConverters.toScala(cs);
    }

    /**
     * This method fetch and return list of Node object for given graph & identifiers
     * @param graphId
     * @param identifiers
     * @return List<Node>
     */
    private static Future<List<Node>> getDataNodes(String graphId, List<String> identifiers) {
        SearchCriteria searchCriteria = new SearchCriteria();
        MetadataCriterion mc = null;
        if (identifiers.size() == 1) {
            mc = MetadataCriterion
                    .create(Arrays.asList(new Filter(SystemProperties.IL_UNIQUE_ID.name(), SearchConditions.OP_EQUAL, identifiers.get(0))));
        } else {
            mc = MetadataCriterion.create(Arrays.asList(new Filter(SystemProperties.IL_UNIQUE_ID.name(), SearchConditions.OP_IN, identifiers)));
        }
        searchCriteria.addMetadata(mc);
        searchCriteria.setCountQuery(false);
        try {
            Future<List<Node>> nodes = SearchAsyncOperations.getNodeByUniqueIds(graphId, searchCriteria);
            return nodes;
        } catch (Exception e) {
            throw new ServerException(GraphEngineErrorCodes.ERR_GRAPH_PROCESSING_ERROR.name(), "Unable To Fetch Nodes From Graph. Exception is: " + e.getMessage());

        }
    }
}

File: ontology-engine/graph-core_2.11/src/main/scala/org/sunbird/graph/validator/NodeValidator.scala
package org.sunbird.graph.validator

import java.util
import java.util.concurrent.CompletionException

import org.sunbird.common.exception.{ResourceNotFoundException, ServerException}
import org.sunbird.graph.common.enums.SystemProperties
import org.sunbird.graph.dac.model.{Filter, MetadataCriterion, Node, SearchConditions, SearchCriteria}
import org.sunbird.graph.exception.GraphErrorCodes
import org.sunbird.graph.service.operation.SearchAsyncOperations

import scala.collection.JavaConversions._
import scala.collection.JavaConverters._

import scala.concurrent.{ExecutionContext, Future}

object NodeValidator {

    def validate(graphId: String, identifiers: util.List[String])(implicit ec: ExecutionContext): Future[util.Map[String, Node]] = {
        val nodes = getDataNodes(graphId, identifiers)

        nodes.map(dataNodes => {
            if (dataNodes.size != identifiers.size) {
                val dbNodeIds = dataNodes.map(node => node.getIdentifier).toList
                val invalidIds = identifiers.toList.filter(id => !dbNodeIds.contains(id))
                throw new ResourceNotFoundException(GraphErrorCodes.ERR_INVALID_NODE.toString, "Node Not Found With Identifier " + invalidIds)
            } else {
                new util.HashMap[String, Node](dataNodes.map(node => node.getIdentifier -> node).toMap.asJava)
            }
        }) recoverWith {
            case e: CompletionException => throw e.getCause
        }
    }

    private def getDataNodes(graphId: String, identifiers: util.List[String]) = {
        val searchCriteria = new SearchCriteria
        val mc = if (identifiers.size == 1)
            MetadataCriterion.create(util.Arrays.asList(new Filter(SystemProperties.IL_UNIQUE_ID.name, SearchConditions.OP_EQUAL, identifiers.get(0))))
        else
            MetadataCriterion.create(util.Arrays.asList(new Filter(SystemProperties.IL_UNIQUE_ID.name, SearchConditions.OP_IN, identifiers), new Filter("status", SearchConditions.OP_NOT_EQUAL, "Retired")))
        searchCriteria.addMetadata(mc)
        searchCriteria.setCountQuery(false)
        try {
            val nodes = SearchAsyncOperations.getNodeByUniqueIds(graphId, searchCriteria)
            nodes
        } catch {
            case e: Exception =>
                throw new ServerException(GraphErrorCodes.ERR_GRAPH_PROCESSING_ERROR.toString, "Unable To Fetch Nodes From Graph. Exception is: " + e.getMessage)
        }
    }
}
